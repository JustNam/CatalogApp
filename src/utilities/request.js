export const callAPI = async (enpoint) => {
  const response = await fetch(enpoint);
  const json = await response.json();
  return json;
};
