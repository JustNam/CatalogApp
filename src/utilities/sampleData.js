export const errorItem = {
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4)',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const itemWithEmptyCategory = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [],
};

export const noDescriptionItem = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const item = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 3,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 8,
      title: 'Ru3',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply',
      id: 9,
      title: 'Ru2',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const category = {
  data: [
    {
      created_on: '06/26/2019, 08:09:32',
      id: 1,
      name: 'Swimming',
      updated_on: '06/26/2019, 08:09:32',
    },
    {
      created_on: '06/26/2019, 08:09:32',
      id: 2,
      name: 'Running',
      updated_on: '06/26/2019, 08:09:32',
    },
  ],
};
