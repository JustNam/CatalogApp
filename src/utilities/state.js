export const initalCategoryState = {
  data: [],
};

export const initalItemState = {
  data: [],
  categoryId: 0,
  currentPage: 0,
  lastPage: 0,
};

export const initalUserState = {
  loggedIn: localStorage.getItem('accessToken') || false,
};