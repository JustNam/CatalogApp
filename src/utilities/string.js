export const convertCamelCaseToSnakeCase = string => (
  string.replace(/[A-Z]/g, match => `_${match.toLowerCase()}`)
);