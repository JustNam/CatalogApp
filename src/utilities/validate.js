export const validateItemTitle = (title) => {
  const { value } = title;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9\s]+$/)) {
    errors.push('Given title contains invalid character');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};

export const validateUsername = (username) => {
  const { value } = username;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9]+$/)) {
    errors.push('Given username contains invalid character');
  }
  if (!value.length > 30) {
    errors.push('Given username contains more than 30 characters');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};