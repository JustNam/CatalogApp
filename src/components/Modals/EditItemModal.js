import React, { Component } from 'react';
import { Modal, Button, Form } from 'react-bootstrap';
import { connect } from 'react-redux';
import { editItemInCategory } from 'actions/item';
import { validateItemTitle } from 'utilities/validate';

export class EditItemModal extends Component {
  constructor(props) {
    super(props);
    const { item } = props;
    this.state = {
      isValidRequest: false,
      title: {
        value: item.title,
        isValid: false,
        errorMessages: [],
      },
      description: {
        value: item.description,
      },
      // Attribute for <input/> element
      titleInputProps: {
        isValid: null,
        isInvalid: null,
      },
      itemId: item.id,
    };
  }

  componentDidMount() {
    // Validate of current title
    this.validateData(this.state.title);
  }

  // Update UI of title input
  updateTitleInputProps = (validity) => {
    this.setState({
      titleInputProps: {
        isValid: validity,
        isInvalid: !validity,
      },
    });
  };

  resetTitleInputProps = () => {
    this.setState({
      titleInputProps: {
        isValid: '',
        isInvalid: '',
      },
    });
  };

  // Validate title of item and then save it to state
  validateData = (title) => {
    const validation = validateItemTitle(title);
    this.setState({
      isValidRequest: validation.isValid,
      title: {
        ...title,
        ...validation,
      },
    });
    this.updateTitleInputProps(validation.isValid);
  };

  // Update title from input and trigger validadtions
  handleTitleChange = (e) => {
    const { title } = this.state;
    title.value = e.target.value;
    if (title.value === '') {
      this.resetTitleInputProps();
    } else {
      this.validateData(title);
    }
  };

  handleDescriptionChange = (e) => {
    if (e.target.value) {
      this.setState({ description: { value: e.target.value } });
    } else {
      this.setState({ description: { value: null } });
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      title,
      description,
      isValidRequest,
      itemId,
    } = this.state;
    // Validate by front-end
    if (isValidRequest) {
      // Edit an item
      this.props
        .editItemInCategory(this.props.currentCategoryId, {
          id: itemId,
          title: title.value.trim(),
          description: description.value,
        })
        .then((response) => {
          // Validate by back-end
          if (response.payload.error) {
            this.setState({
              title: {
                ...title,
                errorMessages: ['The given title already exists'],
              },
            });
            this.updateTitleInputProps(false);
          } else {
            this.props.showSuccessModal();
            this.props.handleClose();
          }
        });
    }
  };

  render() {
    const { title, titleInputProps, description } = this.state;
    return (
      <Modal
        animation={false}
        show={this.props.show}
        onHide={this.props.handleClose}
      >
        <Modal.Header closeButton>
          <Modal.Title>Edit item</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group controlId="formBasicTitle">
              <Form.Label>Title</Form.Label>
              <Form.Control
                name="title"
                required
                onChange={this.handleTitleChange}
                value={title.value}
                {...titleInputProps}
              />
              <Form.Text className="text-muted">
                Title only contain letters, numbers and spaces. Maximum length
                is 30 characters.
              </Form.Text>
              {/* Show error detais */}
              <Form.Control.Feedback type="invalid">
                {title.errorMessages.map((errorMessage, index) => (
                  <div key={index}>{errorMessage}</div>
                ))}
              </Form.Control.Feedback>
            </Form.Group>

            <Form.Group controlId="formBasicPassword">
              <Form.Label>Description</Form.Label>
              <Form.Control
                as="textarea"
                name="description"
                onChange={this.handleDescriptionChange}
                value={description.value}
              />
            </Form.Group>
            <Button type="submit">Edit</Button>
          </Form>
        </Modal.Body>
      </Modal>
    );
  }
}

const mapDispatchToProp = {
  editItemInCategory,
};

export default connect(
  null,
  mapDispatchToProp
)(EditItemModal);
