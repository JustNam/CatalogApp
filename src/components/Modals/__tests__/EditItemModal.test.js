import React from 'react';
import { shallow } from 'enzyme';
import { EditItemModal } from '../EditItemModal';
import { item, errorItem } from '../../../utilities/sampleData';

describe('EditItemModal with valid data', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      item: item.data[0],
      itemId: 1,
      isValidRequest: false,
      title: {
        value: '',
        isValid: false,
        errorMessages: [],
      },
      description: {
        value: '',
      },
      titleInputProps: {
        isValid: null,
        isInvalid: null,
      },
      editItemInCategory: jest.fn(() => Promise.resolve({
        payload: {
          data: '',
        },
      })),
      showSuccessModal: jest.fn(() => {}),
      handleClose: jest.fn(() => {}),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<EditItemModal {...props} />);
    instance = wrapper.instance();
  });
  it('It should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('It should change description in state when description input is changed', () => {
    expect(wrapper.state().isValidRequest).toBe(true);
  });
  it('It should change username in state when username input is changed', () => {
    instance.handleTitleChange({
      target: {
        value: 'test',
      },
    });
    expect(wrapper.state().title.value).toBe('test');
    expect(wrapper.state().titleInputProps.isInvalid).toBe(false);

    instance.handleTitleChange({
      target: {
        value: '',
      },
    });
    expect(wrapper.state().title.value).toBe('');
  });
  it('It should change description in state when description input is changed', () => {
    instance.handleDescriptionChange({
      target: {
        value: 'test',
      },
    });
    expect(wrapper.state().description.value).toBe('test');
    instance.handleDescriptionChange({
      target: {
        value: '',
      },
    });
    expect(wrapper.state().description.value).toBe(null);
  });
  it('It should show success modal if the data is valid', async () => {
    wrapper.setState({
      isValidRequest: true,
    });
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(props.showSuccessModal).toBeCalled();
  });
});

describe('EditItemModal with valid data', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      item: errorItem.data[0],
      isValidRequest: false,
      title: {
        value: '',
        isValid: false,
        errorMessages: [],
      },
      description: {
        value: '',
      },
      titleInputProps: {
        isValid: null,
        isInvalid: null,
      },
      editItemInCategory: jest.fn(() => Promise.resolve({
        payload: {
          error: '1',
        },
      })),
      showSuccessModal: jest.fn(() => {}),
      handleClose: jest.fn(() => {}),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<EditItemModal {...props} />);
    instance = wrapper.instance();
  });
  it('It should show error if the data is considered invalid in back end', async () => {
    wrapper.setState({
      isValidRequest: true,
    });
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(wrapper.state().titleInputProps.isValid).toBe(false);
  });

  it('It should not call API if the data is considered invalid in front end', async () => {
    wrapper.setState({
      isValidRequest: false,
    });
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(props.editItemInCategory).not.toHaveBeenCalled();
  });
});