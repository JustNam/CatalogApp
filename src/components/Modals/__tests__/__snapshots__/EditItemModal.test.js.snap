// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditItemModal with valid data It should render correctly 1`] = `ShallowWrapper {}`;
